viewing_party/party.py

```diff
@@ -1,23 +1,176 @@
+            "genre": genre,
+            "rating": rating
+        }
+        return movie
+    else: 
+        return None
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    wacthlist = user_data["watchlist"]
+
+    for movie in wacthlist:
+        if title == movie["title"]:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    """
+    Calculates the average rating of watched movies.
+
+    Returns:
+    float: The average rating of watched movies, or 0.0 if none are watched.
+    """
+
+    avg_rating = 0.0
+    count = 0
+    sum = 0
+
+    if not user_data["watched"]:
+        return avg_rating
+    
+    for movie in user_data["watched"]:
+        sum += movie["rating"]
+        count += 1
 
+    avg_rating = sum / count
+    return avg_rating
+
+def get_most_watched_genre(user_data):
+    """
+    Determines the most watched genre among watched movies.
+
+    Returns:
+    str: The most watched genre, or None if no movies have been watched.
+    """
+
+    movie_genre = {}
+    count = 0
+    genre = ""
+
+    if not user_data["watched"]:
+        return None
+    
+    for movie in user_data["watched"]:
+        movie_genre[movie["genre"]] = movie_genre.get(movie["genre"], 0) + 1
+        if count < movie_genre[movie["genre"]]:
+            count = movie_genre[movie["genre"]]
+            genre = movie["genre"]
+
+    return genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    """
+    Finds movies that the user has watched but their friends haven't.
+
+    Returns:
+    list: A list of movies watched only by the user.
+    """
+
+    user_watched = user_data.get("watched", [])
+    friends_watched = []
+    user_unique_movies = []
+
+    for friend in user_data.get("friends", []):
+        friends_watched.extend(friend.get("watched", []))
+
+    for movie in user_watched:
+        if movie not in friends_watched and movie not in user_unique_movies:
+            user_unique_movies.append(movie)
+    
+    return user_unique_movies
 
+def get_friends_unique_watched(user_data):
+    """
+    Finds movies that friends have watched but the user hasn't.
+
+    Returns:
+    list: A list of movies watched only by friends.
+    """
+
+    user_watched = user_data.get("watched", [])
+    friends_watched = []
+    friends_unique_movies= []
+
+    for friend in user_data.get("friends", []):
+        friends_watched.extend(friend.get("watched", []))

```

Comment:
Notice that lines 115-116 here repeats exactly what you already wrote on lines 94-95. 

How could you DRY up your code and remove this repetition?

Consider creating a helper function that returns a list of the movies that friends have watched (maybe something like `get_friends_watched_movies`). Then you can invoke that helper function in `get_unique_watched` and here in `get_friends_unique_watched`. 

Not only does this DRY up your code, but it also makes these methods single responsibility. Instead of having to get the friends' movies and then decide if the movie is unique in someway, the logic can just be kept to determining uniqueness.